### PR TITLE
JDBC: Travel Agency

### PR DESCRIPTION
**Pull request by denisdimitrov**
Travel Agency with the functionality to insert, update, display all, display people with certain name, display people in the same city at the same time, truncate a table. The classes Person and Trip only have constructors and nothing else, the methods that are executing queries are in the Travel Agency class. The EGN is primary key in People and foreign key in Trip.
screenshot from 2016-11-21 13-13-12

### CODE REVIEW

File: src/main/java/travelagency/TravelAgency.java [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+        this.connection = connection;
+    }
+
+    /**
+     * Inserts into the table a person with values for name,egn,age and email.
+     *
+     * @param person
+     * @return
+     */
+    public Person insertPerson(Person person) {
```
Comment (EmilGeorgiev) [inline src/main/java/travelagency/TravelAgency.java]: we have duplication of word ```Person```
```java
insertPerson(Person)
``` 

rename method to ```insert(Person)```

Also this method accept Person and return Person what is difference between this two objects

File: src/main/java/travelagency/TravelAgency.java [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+            PreparedStatement preparedStatement = connection.prepareStatement(insert);
+            preparedStatement.setString(1, person.name);
+            preparedStatement.setInt(2, person.egn);
+            preparedStatement.setInt(3, person.age);
+            preparedStatement.setString(4, person.email);
+            preparedStatement.executeUpdate();
+        } catch (SQLException e) {
+            e.printStackTrace();
+        }
+        return null;
```
Comment (EmilGeorgiev) [inline src/main/java/travelagency/TravelAgency.java]: it is bad to return ```null```

File: src/main/java/travelagency/TravelAgency.java [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+     */
+    public Trip insertTrip(Trip trip) {
+        try {
+            String insert = "insert into trip values (?,?,?,?)";
+            PreparedStatement preparedStatement = connection.prepareStatement(insert);
+            preparedStatement.setInt(1, trip.egn);
+            preparedStatement.setInt(2, trip.arrival);
+            preparedStatement.setInt(3, trip.departure);
+            preparedStatement.setString(4, trip.city);
+            preparedStatement.executeUpdate();
```
Comment (EmilGeorgiev) [inline src/main/java/travelagency/TravelAgency.java]: ```preparedStatement.set...``` is duplicated 20 times. Ask your colleague @AlexanderVladimirov9090 to give you Implmentation of Datastore and use it.

File: src/main/java/travelagency/TravelAgency.java (hunk 1 above)
Comment (denisdimitrov) [inline src/main/java/travelagency/TravelAgency.java]: At first both of my methods insertPerson and insertTrip were with name insert and that's why. Then I decided to make them voids but I forgot. I will fix it as soon as possible.

File: src/main/java/travelagency/TravelAgency.java (hunk 2 above)
Comment (denisdimitrov) [inline src/main/java/travelagency/TravelAgency.java]: I will avoid it in the future.

**Comment by EmilGeorgiev** [comment]
here an example of custom Rule: http://www.singhajit.com/junit-rules/

**Comment by denisdimitrov** [comment]
Used DataStore to reduce the use or PreparedStatements for less code and better readability. Also added tests for all of the methods. The fields in Trip and Person are now immutable. Also used the custom rule to create and drop tables at the start and end of each test.

File: src/main/java/core/PeopleFinder.java [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package core;
+
+import java.util.Date;
+import java.util.List;
+
+/**
+ * @author Denis Dimitrov <denis.k.dimitrov@gmail.com>
+ */
+public interface PeopleFinder {
+    List<Person> showPeople();
```
Comment (EmilGeorgiev) [inline src/main/java/core/PeopleFinder.java]: fetchAll

File: src/main/java/core/PeopleFinder.java [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+    List<Person> showPeopleByName(String name);
+    List<Person> showPeopleSameLocation(int date, String city);
```
Comment (EmilGeorgiev) [inline src/main/java/core/PeopleFinder.java]: Name people is duplicated in Name of interface and name of methods. When name of interface is some concrete as ```PeopleFilder``` we can omit word People in name of methods

**Comment by denisdimitrov** [comment]
The tests are working properly now with before and after annotations. Also I removed the custom rule because it was unnecessary. Also the database can now be chosen from the constructor.

File: src/test/java/persistence/PersistencePeopleTest.java [hunk 6]
```diff
@@ -44,56 +43,56 @@ public void tearDown() throws Exception {
     }
 
     @Test
-    public void insertPerson() throws Exception {
+    public void registerPerson() throws Exception {
         Person person = new Person("Denis", 1231231231, 22, "asd@gmail.com");
-        people.insert(person);
+        people.register(person);
         List<Person> list = new ArrayList<>();
         list.add(person);
```
Comment (EmilGeorgiev) [inline src/test/java/persistence/PersistencePeopleTest.java]: use Lists.newArrayList(person) https://google.github.io/guava/releases/snapshot/api/docs/com/google/common/collect/Lists.html

**Comment by denisdimitrov** [comment]
The lists in the tests are now from guava lib.

File: src/test/java/persistence/PersistencePeopleTest.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    @After
+    public void tearDown() throws Exception {
+        String dropTablePeople = "drop table people";
+        connection.get().prepareStatement(dropTablePeople).executeUpdate();
+        String dropTableTrip = "drop table trip";
+        connection.get().prepareStatement(dropTableTrip).executeUpdate();
+    }
+
+    @Test
+    public void registerPerson() throws Exception {
```
Comment (EmilGeorgiev) [inline src/test/java/persistence/PersistencePeopleTest.java]: this test is unnecessary, because we use register in all other tests

Comment (denisdimitrov) [inline src/test/java/persistence/PersistencePeopleTest.java]: I will remove the first tests from both Person and Trip.

**Pull request merged**